lib/SILOptimizer/Transforms/ArrayElementValuePropagation.cpp

```diff
@@ -258,28 +320,59 @@ class ArrayElementPropagation : public SILFunctionTransform {
 
   void run() override {
     auto &Fn = *getFunction();
+    auto &M = Fn.getModule();
 
     bool Changed = false;
 
     // Propagate the elements an of array value to its users.
-    SmallVector<std::pair<ApplyInst *, SILValue>, 16> ValueReplacements;
+    SmallVector<std::pair<ApplyInst *, SILValue>, 16> GetElementReplacements;
+    SmallVector<std::pair<ApplyInst *, llvm::SmallVector<SILValue, 4>>, 4> AppendContentsOfReplacements;
     for (auto &BB :Fn) {
       for (auto &Inst : BB) {
         if (auto *Apply = dyn_cast<ApplyInst>(&Inst))
           Changed |=
-              ArrayAllocation::findValueReplacements(Apply, ValueReplacements);
+              ArrayAllocation::findReplacements(Apply, GetElementReplacements, AppendContentsOfReplacements);
       }
     }
-    DEBUG(if (Changed) {
+
+    DEBUG(if (!GetElementReplacements.empty()) {
       llvm::dbgs() << "Array elements replaced in " << Fn.getName() << " ("
-                   << ValueReplacements.size() << ")\n";
+                   << GetElementReplacements.size() << ")\n";
     });
     // Perform the actual replacement of the get_element call by its value.
-    for (auto &Repl : ValueReplacements) {
+    for (auto &Repl : GetElementReplacements) {
       ArraySemanticsCall GetElement(Repl.first);
       GetElement.replaceByValue(Repl.second);
     }
 
+    DEBUG(if (!AppendContentsOfReplacements.empty()) {
+      llvm::dbgs() << "Array append contentsOf calls replaced in " << Fn.getName() << " ("
+                   << AppendContentsOfReplacements.size() << ")\n";
+    });
+    // Perform the actual replacement of the append_contentsOf call by its value.
+    if (!AppendContentsOfReplacements.empty()) {
+      auto *appendFnDecl = lookupAppendFnDecl();
+      assert(appendFnDecl && "Array.append(Element) declaration must exist");
+
+      auto mangled = SILDeclRef(appendFnDecl, SILDeclRef::Kind::Func).mangle();
+      auto *appendFn = M.hasFunction(mangled, SILLinkage::PublicExternal);
+      assert(appendFn && "Array.append(Element) function must exist");
```

Comment:
I think it's better not to rely on the existence of the append function. It's better to bail instead of assert here.